File: isthmus/src/main/java/io/substrait/isthmus/expression/FunctionMappings.java
package io.substrait.isthmus.expression;

import com.google.common.collect.ImmutableList;
import io.substrait.isthmus.AggregateFunctions;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import org.apache.calcite.sql.SqlOperator;
import org.apache.calcite.sql.fun.SqlLibraryOperators;
import org.apache.calcite.sql.fun.SqlStdOperatorTable;

/**
 * Defines static mappings between Calcite {@link SqlOperator} signatures and Substrait base
 * function names, including scalar, aggregate, and window function signatures.
 *
 * <p>Also provides type-based resolvers to disambiguate operators by output type.
 */
public class FunctionMappings {

  /** Scalar operator signatures mapped to Substrait function names. */
  public static final ImmutableList<Sig> SCALAR_SIGS =
      ImmutableList.<Sig>builder()
          .add(
              s(SqlStdOperatorTable.PLUS, "add"),
              s(SqlStdOperatorTable.MINUS, "subtract"),
              s(SqlStdOperatorTable.UNARY_MINUS, "negate"),
              s(SqlStdOperatorTable.MULTIPLY, "multiply"),
              s(SqlStdOperatorTable.DIVIDE, "divide"),
              s(SqlStdOperatorTable.ABS, "abs"),
              s(SqlStdOperatorTable.MOD, "modulus"),
              s(SqlStdOperatorTable.POWER, "power"),
              s(SqlStdOperatorTable.EXP, "exp"),
              s(SqlStdOperatorTable.SIN, "sin"),
              s(SqlStdOperatorTable.COS, "cos"),
              s(SqlStdOperatorTable.TAN, "tan"),
              s(SqlStdOperatorTable.ASIN, "asin"),
              s(SqlStdOperatorTable.ACOS, "acos"),
              s(SqlStdOperatorTable.ATAN, "atan"),
              s(SqlStdOperatorTable.ATAN2, "atan2"),
              s(SqlStdOperatorTable.SIGN, "sign"),
              s(SqlStdOperatorTable.LOG10, "log10"),
              s(SqlStdOperatorTable.LN, "ln"),
              s(SqlStdOperatorTable.AND),
              s(SqlStdOperatorTable.OR),
              s(SqlStdOperatorTable.NOT),
              s(SqlStdOperatorTable.LESS_THAN, "lt"),
              s(SqlStdOperatorTable.LESS_THAN_OR_EQUAL, "lte"),
              s(SqlStdOperatorTable.GREATER_THAN, "gt"),
              s(SqlStdOperatorTable.GREATER_THAN_OR_EQUAL, "gte"),
              s(SqlStdOperatorTable.EQUALS, "equal"),
              s(SqlStdOperatorTable.BIT_XOR, "xor"),
              s(SqlStdOperatorTable.IS_NULL, "is_null"),
              s(SqlStdOperatorTable.IS_NOT_NULL, "is_not_null"),
              s(SqlStdOperatorTable.NOT_EQUALS, "not_equal"),
              s(SqlStdOperatorTable.MINUS_DATE, "subtract"),
              s(SqlStdOperatorTable.DATETIME_PLUS, "add"),
              s(SqlStdOperatorTable.EXTRACT, "extract"),
              s(SqlStdOperatorTable.CEIL, "ceil"),
              s(SqlStdOperatorTable.FLOOR, "floor"),
              s(SqlStdOperatorTable.ROUND, "round"),
              s(SqlStdOperatorTable.LIKE),
              s(SqlStdOperatorTable.REPLACE, "replace"),
              s(SqlStdOperatorTable.SUBSTRING, "substring"),
              s(SqlStdOperatorTable.CONCAT, "concat"),
              s(SqlStdOperatorTable.CHAR_LENGTH, "char_length"),
              s(SqlStdOperatorTable.LOWER, "lower"),
              s(SqlStdOperatorTable.UPPER, "upper"),
              s(SqlStdOperatorTable.BETWEEN),
              s(SqlStdOperatorTable.IS_NOT_DISTINCT_FROM, "is_not_distinct_from"),
              s(SqlStdOperatorTable.COALESCE, "coalesce"),
              s(SqlStdOperatorTable.TRIM, "trim"),
              s(SqlStdOperatorTable.TRIM, "ltrim"),
              s(SqlStdOperatorTable.TRIM, "rtrim"),
              s(SqlStdOperatorTable.SQRT, "sqrt"),
              s(SqlLibraryOperators.SINH, "sinh"),
              s(SqlLibraryOperators.TANH, "tanh"),
              s(SqlLibraryOperators.COSH, "cosh"),
              s(SqlLibraryOperators.ASINH, "asinh"),
              s(SqlLibraryOperators.ATANH, "atanh"),
              s(SqlLibraryOperators.ACOSH, "acosh"),
              s(SqlStdOperatorTable.BITNOT, "bitwise_not"),
              s(SqlStdOperatorTable.BITOR, "bitwise_or"),
              s(SqlStdOperatorTable.BITAND, "bitwise_and"),
              s(SqlStdOperatorTable.BITXOR, "bitwise_xor"),
              s(SqlStdOperatorTable.RADIANS, "radians"),
              s(SqlStdOperatorTable.DEGREES, "degrees"),
              s(SqlLibraryOperators.FACTORIAL, "factorial"),
              s(SqlStdOperatorTable.IS_TRUE, "is_true"),
              s(SqlStdOperatorTable.IS_FALSE, "is_false"),
              s(SqlStdOperatorTable.IS_NOT_TRUE, "is_not_true"),
              s(SqlStdOperatorTable.IS_NOT_FALSE, "is_not_false"),
              s(SqlStdOperatorTable.IS_DISTINCT_FROM, "is_distinct_from"),
              s(SqlLibraryOperators.LOG2, "log2"),
              s(SqlLibraryOperators.LEAST, "least"),
              s(SqlLibraryOperators.GREATEST, "greatest"),
              s(SqlStdOperatorTable.BIT_LEFT_SHIFT, "shift_left"),
              s(SqlStdOperatorTable.LEFTSHIFT, "shift_left"),
              s(SqlLibraryOperators.STARTS_WITH, "starts_with"),
              s(SqlLibraryOperators.ENDS_WITH, "ends_with"),
              s(SqlLibraryOperators.CONTAINS_SUBSTR, "contains"),
              s(SqlStdOperatorTable.POSITION, "strpos"),
              s(SqlLibraryOperators.LEFT, "left"),
              s(SqlLibraryOperators.RIGHT, "right"),
              s(SqlLibraryOperators.LPAD, "lpad"),
              s(SqlLibraryOperators.RPAD, "rpad"))
          .build();

  /** Aggregate operator signatures mapped to Substrait function names. */
  public static final ImmutableList<Sig> AGGREGATE_SIGS =
      ImmutableList.<Sig>builder()
          .add(
              s(AggregateFunctions.MIN, "min"),
              s(AggregateFunctions.MAX, "max"),
              s(AggregateFunctions.SUM, "sum"),
              s(AggregateFunctions.SUM0, "sum0"),
              s(SqlStdOperatorTable.COUNT, "count"),
              s(SqlStdOperatorTable.APPROX_COUNT_DISTINCT, "approx_count_distinct"),
              s(AggregateFunctions.AVG, "avg"))
          .build();

  /** Window function signatures (including supported aggregates) mapped to Substrait names. */
  public static final ImmutableList<Sig> WINDOW_SIGS =
      ImmutableList.<Sig>builder()
          .add(
              s(SqlStdOperatorTable.ROW_NUMBER, "row_number"),
              s(SqlStdOperatorTable.LAG, "lag"),
              s(SqlStdOperatorTable.LEAD, "lead"),
              s(SqlStdOperatorTable.RANK, "rank"),
              s(SqlStdOperatorTable.DENSE_RANK, "dense_rank"),
              s(SqlStdOperatorTable.PERCENT_RANK, "percent_rank"),
              s(SqlStdOperatorTable.CUME_DIST, "cume_dist"),
              s(SqlStdOperatorTable.NTILE, "ntile"),
              s(SqlStdOperatorTable.FIRST_VALUE, "first_value"),
              s(SqlStdOperatorTable.LAST_VALUE, "last_value"),
              s(SqlStdOperatorTable.NTH_VALUE, "nth_value"))
          // Aggregate Functions can be used in Windows
          .addAll(AGGREGATE_SIGS)
          .build();

  /** Type-based resolvers to disambiguate Calcite operators by expected output type. */
  public static final Map<SqlOperator, TypeBasedResolver> OPERATOR_RESOLVER =
      Map.of(
          SqlStdOperatorTable.PLUS,
          resolver(
              SqlStdOperatorTable.PLUS, Set.of("i8", "i16", "i32", "i64", "fp32", "fp64", "dec")),
          SqlStdOperatorTable.DATETIME_PLUS,
          resolver(SqlStdOperatorTable.PLUS, Set.of("date", "time", "timestamp")),
          SqlStdOperatorTable.MINUS,
          resolver(
              SqlStdOperatorTable.MINUS, Set.of("i8", "i16", "i32", "i64", "fp32", "fp64", "dec")),
          SqlStdOperatorTable.MINUS_DATE,
          resolver(SqlStdOperatorTable.MINUS_DATE, Set.of("date", "timestamp_tz", "timestamp")),
          SqlStdOperatorTable.BIT_LEFT_SHIFT,
          resolver(SqlStdOperatorTable.BIT_LEFT_SHIFT, Set.of("i8", "i16", "i32", "i64")));

  /**
   * Prints all scalar signatures (for quick inspection).
   *
   * @param args CLI arguments (unused)
   */
  public static void main(String[] args) {
    SCALAR_SIGS.forEach(System.out::println);
  }

  /**
   * Creates a signature mapping entry.
   *
   * @param operator the Calcite operator
   * @param substraitName the Substrait canonical function name
   * @return a {@link Sig} instance
   */
  public static Sig s(SqlOperator operator, String substraitName) {
    return new Sig(operator, substraitName.toLowerCase(Locale.ROOT));
  }

  /**
   * Creates a signature mapping entry using the operator's own (lowercased) name.
   *
   * @param operator the Calcite operator
   * @return a {@link Sig} instance
   */
  public static Sig s(SqlOperator operator) {
    return s(operator, operator.getName().toLowerCase(Locale.ROOT));
  }

  /** Simple signature tuple of operator to Substrait name. */
  public static class Sig {

    /** SqlOperator. */
    public final SqlOperator operator;

    /** Name. */
    public final String name;

    /**
     * Constructs a signature entry.
     *
     * @param operator the Calcite operator
     * @param name the Substrait function name
     */
    public Sig(final SqlOperator operator, final String name) {
      this.operator = operator;
      this.name = name;
    }

    /**
     * Returns the Substrait function name.
     *
     * @return the Substrait name
     */
    public String name() {
      return name;
    }

    /**
     * Returns the Calcite operator.
     *
     * @return the operator
     */
    public SqlOperator operator() {
      return operator;
    }
  }

  /**
   * Creates a type-based resolver for an operator.
   *
   * @param operator the Calcite operator
   * @param outTypes the set of allowed output type strings (Substrait)
   * @return a {@link TypeBasedResolver}
   */
  public static TypeBasedResolver resolver(SqlOperator operator, Set<String> outTypes) {
    return new TypeBasedResolver(operator, outTypes);
  }

  /** Disambiguates operators based on expected output type strings. */
  public static class TypeBasedResolver {

    /** SqlOperator. */
    public final SqlOperator operator;

    /** Types. */
    public final Set<String> types;

    /**
     * Constructs a resolver.
     *
     * @param operator the Calcite operator
     * @param types allowed output type strings
     */
    public TypeBasedResolver(final SqlOperator operator, final Set<String> types) {
      this.operator = operator;
      this.types = types;
    }

    /**
     * Returns the operator this resolver applies to.
     *
     * @return the operator
     */
    public SqlOperator operator() {
      return operator;
    }

    /**
     * Returns the allowed output type strings.
     *
     * @return set of type strings
     */
    public Set<String> types() {
      return types;
    }
  }
}

File: spark/src/main/scala/io/substrait/spark/expression/FunctionMappings.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.substrait.spark.expression

import io.substrait.spark.utils.Util

import org.apache.spark.sql.catalyst.analysis.FunctionRegistryBase
import org.apache.spark.sql.catalyst.expressions._
import org.apache.spark.sql.catalyst.expressions.aggregate._
import org.apache.spark.sql.types.{DayTimeIntervalType, IntegerType}

import scala.reflect.ClassTag

trait Sig {
  def name: String
  def expClass: Class[_]
  def makeCall(args: Seq[Expression]): Expression
}

case class GenericSig(expClass: Class[_], name: String, builder: Seq[Expression] => Expression)
  extends Sig {
  override def makeCall(args: Seq[Expression]): Expression = {
    builder(args)
  }
}

case class SpecialSig(
    expClass: Class[_],
    name: String,
    key: Option[String],
    builder: Seq[Expression] => Expression)
  extends Sig {
  override def makeCall(args: Seq[Expression]): Expression = {
    builder(args)
  }
}

object DateFunction {
  def unapply(e: Expression): Option[Seq[Expression]] = e match {
    case DateAdd(startDate, days) => Some(Seq(startDate, days))
    // The following map to the Substrait `extract` function.
    case Year(date) => Some(Seq(Enum("YEAR"), date))
    case Quarter(date) => Some(Seq(Enum("QUARTER"), Enum("ONE"), date))
    case Month(date) => Some(Seq(Enum("MONTH"), Enum("ONE"), date))
    case DayOfMonth(date) => Some(Seq(Enum("DAY"), Enum("ONE"), date))
    case _ => None
  }

  def unapply(name_args: (String, Seq[Expression])): Option[Expression] = name_args match {
    case ("add:date_i32", Seq(startDate, days)) => Some(DateAdd(startDate, days))
    case ("extract", Seq(Enum("YEAR"), date)) => Some(Year(date))
    case ("extract", Seq(Enum("QUARTER"), Enum("ONE"), date)) => Some(Quarter(date))
    case ("extract", Seq(Enum("MONTH"), Enum("ONE"), date)) => Some(Month(date))
    case ("extract", Seq(Enum("DAY"), Enum("ONE"), date)) => Some(DayOfMonth(date))
    case _ => None
  }
}

object TrimFunction {
  def unapply(e: Expression): Option[Seq[Expression]] = e match {
    case StringTrim(srcStr, trimStr) => Some(Seq(srcStr) ++ trimStr)
    case StringTrimLeft(srcStr, trimStr) => Some(Seq(srcStr) ++ trimStr)
    case StringTrimRight(srcStr, trimStr) => Some(Seq(srcStr) ++ trimStr)
    case _ => None
  }

  def unapply(name_args: (String, Seq[Expression])): Option[Expression] = name_args match {
    case ("trim", Seq(srcStr)) => Some(StringTrim(srcStr))
    case ("trim", Seq(srcStr, trimStr)) => Some(StringTrim(srcStr, trimStr))
    case ("ltrim", Seq(srcStr)) => Some(StringTrimLeft(srcStr))
    case ("ltrim", Seq(srcStr, trimStr)) => Some(StringTrimLeft(srcStr, trimStr))
    case ("rtrim", Seq(srcStr)) => Some(StringTrimRight(srcStr))
    case ("rtrim", Seq(srcStr, trimStr)) => Some(StringTrimRight(srcStr, trimStr))
    case _ => None
  }
}

class FunctionMappings {

  private def s[T <: Expression: ClassTag](name: String): GenericSig = {
    val builder = FunctionRegistryBase.build[T](name, None)._2
    GenericSig(scala.reflect.classTag[T].runtimeClass, name, builder)
  }

  private def ss[T <: Expression: ClassTag](signature: String): SpecialSig = {
    val (name, key) = if (signature.contains(":")) {
      (signature.split(':').head, Some(signature))
    } else {
      (signature, None)
    }
    val builder = (args: Seq[Expression]) =>
      (signature, args) match {
        case DateFunction(expr) => expr
        case TrimFunction(expr) => expr
        case _ =>
          throw new UnsupportedOperationException(s"Cannot convert $signature with arguments $args")
      }
    SpecialSig(scala.reflect.classTag[T].runtimeClass, name, key, builder)
  }

  val SCALAR_SIGS: Seq[Sig] = Seq(
    s[Add]("add"),
    s[Subtract]("subtract"),
    s[Multiply]("multiply"),
    s[Divide]("divide"),
    s[Abs]("abs"),
    s[Remainder]("modulus"),
    s[Round]("round"),
    s[Floor]("floor"),
    s[Ceil]("ceil"),
    s[Pow]("power"),
    s[Exp]("exp"),
    s[Sqrt]("sqrt"),
    s[Sin]("sin"),
    s[Cos]("cos"),
    s[Tan]("tan"),
    s[Asin]("asin"),
    s[Acos]("acos"),
    s[Atan]("atan"),
    s[Atan2]("atan2"),
    s[Sinh]("sinh"),
    s[Cosh]("cosh"),
    s[Tanh]("tanh"),
    s[Asinh]("asinh"),
    s[Acosh]("acosh"),
    s[Atanh]("atanh"),
    s[Log]("ln"),
    s[Log10]("log10"),
    s[And]("and"),
    s[Or]("or"),
    s[Not]("not"),
    s[LessThan]("lt"),
    s[LessThanOrEqual]("lte"),
    s[GreaterThan]("gt"),
    s[GreaterThanOrEqual]("gte"),
    s[EqualTo]("equal"),
    s[EqualNullSafe]("is_not_distinct_from"),
    s[IsNull]("is_null"),
    s[IsNotNull]("is_not_null"),
    s[EndsWith]("ends_with"),
    s[Like]("like"),
    s[Contains]("contains"),
    s[StartsWith]("starts_with"),
    s[Substring]("substring"),
    s[Upper]("upper"),
    s[Lower]("lower"),
    s[Concat]("concat"),
    s[Coalesce]("coalesce"),
    s[ShiftLeft]("shift_left"),
    s[ShiftRight]("shift_right"),
    s[ShiftRightUnsigned]("shift_right_unsigned"),
    s[BitwiseAnd]("bitwise_and"),
    s[BitwiseOr]("bitwise_or"),
    s[BitwiseXor]("bitwise_xor"),

    // trim functions require special handling
    ss[StringTrim]("trim"),
    ss[StringTrimLeft]("ltrim"),
    ss[StringTrimRight]("rtrim"),

    // date/time functions require special handling
    ss[DateAdd]("add:date_i32"),
    ss[Year]("extract"),
    ss[Quarter]("extract"),
    ss[Month]("extract"),
    ss[DayOfMonth]("extract")
  )

  val AGGREGATE_SIGS: Seq[Sig] = Seq(
    s[Sum]("sum"),
    s[Average]("avg"),
    s[Count]("count"),
    s[Min]("min"),
    s[Max]("max"),
    s[First]("any_value"),
    s[HyperLogLogPlusPlus]("approx_count_distinct"),
    s[StddevSamp]("std_dev")
  )

  val WINDOW_SIGS: Seq[Sig] = Seq(
    s[RowNumber]("row_number"),
    s[Rank]("rank"),
    s[DenseRank]("dense_rank"),
    s[PercentRank]("percent_rank"),
    s[CumeDist]("cume_dist"),
    s[NTile]("ntile"),
    s[Lead]("lead"),
    s[Lag]("lag"),
    s[NthValue]("nth_value")
  )
}

object FunctionMappings extends FunctionMappings
